[API] Data Movement Support Enhancement

Hecmay opened the pull request:
In this PR, we introduced same new features to facilitate the p2p and many-to-many data movement patterns. All examples of using `.to` using new features are available here: https://github.com/cornell-zhang/heterocl/blob/v0.3/tests/test_schedule_stream.py

1. **Support data movement from any step in the program**. A tensor can be modified in different stages in the program, we can select which modified version to be moved to the target device. For example in the following case, Tensor A is moved to host after being modified in the `U1` stage.

```python
def kernel(A):
    hcl.update(A, lambda *args: A[args] + 2, "U1")
    hcl.update(A, lambda *args: A[args] * 3, "U2")

s.to(A, target.xcel)
s.to(kernel.U1.A, target.host)
```

2. **Data movement from different levels**. As mentioned in the CAPA slides, we can combine the tiling with data movement to enable data transfer in finer granularity.

```python
def kernel(A):
    hcl.update(A, lambda *args: A[args] * 2, "U1")

xo, xi = s[kernel.U1].split(kernel.U1.axis[1], factor=3)
s.to(kernel.U1, target.xcel, axis=xo)
```

3. **Fork and Join**: Ideally the forked statements (e.g. sub-loop body) are supposed to be executed in parallel with a dispatcher dynamically scheduling the tasks, and a collector write the result to buffers in the right order when they are joined. Implementing such a functionality in IR pass might not be a good choice (can be of too much work, also not performant or area efficient), also in the network processing applications, most packets are preferred to be implemented as a data struct to save the information, e.g. valid signal, virtual channel and payload, which makes it much more challenging to implement. I think it can be good if we can use some IPs to implement the router: #170 .

``` python
def kernel(A, B):
    # broadcasting 
    C = hcl.compute(A.shape, lambda *args: A[args] + 1, "C")
    D = hcl.compute(A.shape, lambda *args: A[args] * 2, "D")

    hcl.update(B, lambda *args: C[args] + 1, "U1")
    hcl.update(B, lambda *args: C[args] * 2, "U2")

s.fork(A, [kernel.C, kernel.D])
s.join([kernel.U1, kernel.U2], B)
```

4. **Enhanced the support for self feedback loop streaming**. Example of Systolic Array with streaming (where K is time axis and X/Y are spatial axes)

```python
    def kernel(A, B, O):

        stream_A = hcl.compute((m, k-1), lambda *args: 0, "stream_A")
        stream_B = hcl.compute((k-1, n), lambda *args: 0, "stream_B")

        def update(k, y, x):
            last = hcl.scalar(
                hcl.select(k==0, 0, O[y, x]), "last")

            stream_A[y, x] = hcl.select(x>0, stream_A[y, x-1], A[y, k])
            stream_B[y, x] = hcl.select(y>0, stream_B[y-1, x], B[k, x])
            O[y, x] = last.v + stream_A[y, x] * stream_B[y, x]

        hcl.mutate((m, dim_y, dim_x), 
            lambda k, y, x: update(k, y, x), name="update")

    s = hcl.create_schedule([A, B, output], kernel)
    s.to(kernel.udpate.stream_A, s[kernel.update])
```
 In the code snippet shown above. The `stream_A` and `stream_B` are local copies of the input on each PE. To stream the data and circumvent the on-chip memory access, we can apply the feedback streaming in the loop to stream the data `stream_A` to the loop it resides in. 

5. **LLVM JIT Simulation for Streaming** (WIP): this part is mainly done by Sean. I will port the LLVM JIT engine to support streaming using multithreading support.

```python
A = hcl.placeholder((10,10))

def kernel(A):
    B = hcl,compute((10,10), lambda *args: A[args] + 1, "B")
    C = hcl,compute((10,10), lambda *args: B[args] + 1, "C")

s = hcl.create_schedule([A], kernel)
# the two loops will be executed in parallel in LLVM simulation 
s.to(kernel.B, kernel.C)
```

6. **One to many** (ideally the producer should check the validity of the consumers before sending data, However in this case we assume all the consumers are always available). 
```python
def kernel(A):
    B = hcl.compute(A.shape, lambda *args: A[args] * 2)
    C = hcl.compute(A.shape, lambda *args: A[args] + 1)

s.fork(A, [kernel.B, kernel.C])
```
By applying fork schedule, the shared input will create two separate streaming channels to send data to the consumers concurrently, reducing the tension of competing read ports of the same memory block (tensor A in this case).

```shell
for ...
    temp = A [args]
    channel1.write(temp)
    channel2.write(temp)

for ...
    B[args] = 2 * channel1.read()
    C[args] = 1 + channel2.read()
```


7. **Many to one** (We create a simple collector in this case, all values can be written in parallel into the streaming channels and will be consumed sequentially by the collector)
```python
def kernel(A):
    hcl.update(C, lambda i, j: A[i,j] + 1, "s1")
    hcl.update(C, lambda i, j: B[i,j] * 2, "s2")
    return hcl.compute(C.shape, lambda *args: C[args] + 3, "ret")

s.join([kernel.s1.C, kernel.s2.C], kernel.ret.C)
```
In this example we have a destination stage for values comping from different producers to be joined. If the destination is absent, a new collector stage will be created to receive and store the data in the correct order.


```shell
// producer 1
for ...
    channel1.write(A[i,j]+1)

// producer 2
for ...
    channel2.write(B[i,j]*2)

// consumer (collector)
while (!channel.empty())
    index, val = channel.read()
    if (condition) C[index] = val
```

8. **Moving data to specific media**: The default `.to` API moves data to main memory cells used on the accelerators. E.g. global DDR attached to FPGA acceleration card. Aside from off-chip DRAM,  HeteroCL also supports moving data to high bandwidth memory and PLRAM on modern FPGAs.  Example to move data to HBM/PLRAM assigned to different channels.

```python
# assign tensor A to channel 0 of HBM, and B to 1
# p.xcel is an FPGA card with HBM equipped.
s.to(A, p.xcel.hbm[0])
s.to(B, p.xcel.hbm[1])
```

9. **Compute Unit Duplication**: Here is an example to perform compute unit duplication on a kernel. Firstly we extract the kernel function based on the boundary information of the DFG (i.e. the data placement information). The kernel function may have multiple stages inside. All the stages inside are combined into a single stage (i.e. the `s[kernel]` stage) using `compute_at`. 

With the combined stage (which is basically a nested loop describing all the computation on kernel), we can perform CU duplication by splitting the outermost loop. The splitting factor here is corresponding to the duplication factor, and we map each split segment into a compute unit.
```python
A_, B_ = s.to([A, B], p.xcel.hbm)
E_ = s.to(kernel.E, p.host)

# we can wrap the following primitives into a high level API
# 1. get the kernel function based on boundary.
kernel = s.subgraph(inputs=[A_, B_], output=[E_])

# 2. split the kernel stage and parallelize the outer loop. 
xo, xi = s[kernel].split(kernel.axis[0], factor=2)
s[kernel].parallel(xo)

# 3. reshape and partition off-chip buffer in certain dimension
s.reshape(A_, (10, 2, 5))
s.reshape(B_, (10, 2, 5))
s.partition([A_, B_], dim=2)
```

Hecmay commented:
Some other updates:

4. **Enhanced the support for self feedback loop streaming**. Example of Systolic Array with streaming (where K is time axis and X/Y are spatial axes)

```python
    def kernel(A, B, O):

        stream_A = hcl.compute((m, k-1), lambda *args: 0, "stream_A")
        stream_B = hcl.compute((k-1, n), lambda *args: 0, "stream_B")

        def update(k, y, x):
            last = hcl.scalar(
                hcl.select(k==0, 0, O[y, x]), "last")

            stream_A[y, x] = hcl.select(x>0, stream_A[y, x-1], A[y, k])
            stream_B[y, x] = hcl.select(y>0, stream_B[y-1, x], B[k, x])
            O[y, x] = last.v + stream_A[y, x] * stream_B[y, x]

        hcl.mutate((m, dim_y, dim_x), 
            lambda k, y, x: update(k, y, x), name="update")

    s = hcl.create_schedule([A, B, output], kernel)
    s.to(kernel.udpate.stream_A, s[kernel.update])
```
 In the code snippet shown above. The `stream_A` and `stream_B` are local copies of the input on each PE. To stream the data and circumvent the on-chip memory access, we can apply the feedback streaming in the loop to stream the data `stream_A` to the loop it resides in. 

5. **LLVM JIT Simulation for Streaming** (WIP): this part is mainly done by Sean. I will port the LLVM JIT engine to support streaming using multithreading support.

Hecmay commented:
New supports for many-to-many communication pattern: this is a simple prototyping with any support for flow control or path routing. Some simple examples:

1. **One to many** (ideally the producer should check the validity of the consumers before sending data, However in this case we assume all the consumers are always available). 
```python
def kernel(A):
    B = hcl.compute(A.shape, lambda *args: A[args] * 2)
    C = hcl.compute(A.shape, lambda *args: A[args] + 1)

s.fork(A, [kernel.B, kernel.C])
```
By applying fork schedule, the shared input will create two separate streaming channels to send data to the consumers concurrently, reducing the tension of competing read ports of the same memory block (tensor A in this case).

```shell
for ...
    temp = A [args]
    channel1.write(temp)
    channel2.write(temp)

for ...
    B[args] = 2 * channel1.read()
    C[args] = 1 + channel2.read()
```

Another case is to fork the statement instances under a loop, where `fork` schedule basically unrolls the inner loop and annotates unrolled loop body with `dataflow` attribute.
```python
def update(y, x):
    A[y, x] = y * x
    B[y, x] = y + x

hcl.mutate((10,10), update, name="update")
s.fork(kernel.update, axis=1, factor=2)
```


2. **Many to one** (We create a simple collector in this case, all values can be written in parallel into the streaming channels and will be consumed sequentially by the collector)

zhangzhiru commented:
fork and join are usually used to described parallel compute patterns. Can we just stick to .to() instead?

Hecmay commented:
> fork and join are usually used to described parallel compute patterns. Can we just stick to .to() instead?

Sure. These two APIs are basically wrapper functions of `.to`. All of the connected components will run in a dataflow fashion. It's in parallel but the components are not identical, so it should be good idea to use `to`.

Hecmay commented:
Here is an example to perform compute unit duplication on a kernel. Firstly we extract the kernel function based on the boundary information of the DFG.
```python
A_, B_ = s.to([A, B], p.xcel, if=p.xcel.hbm)
E_ = s.to(kernel.E, p.host)

# we can wrap the following primitives into a high level API
# 1. get the kernel function based on boundary. all the stages in the kernel 
#.   are combined into a single stage using compute_at
kernel = s.subgraph(inputs=[A_, B_], output=[E_])

# 2. split the kernel stage and parallelize the outer loop. The splitting factor 
#    here is the duplication factor for the compute units
xo, xi = s[kernel].split(kernel.axis[0], factor=2)
s[kernel].parallel(xo)

# 3. reshape and partition off-chip buffer in certain dimension
s.reshape(A, (10, 2, 5))
s.reshape(B, (10, 2, 5))
s.partition([A, B], dim=2)
```

seanlatias commented on python/heterocl/tvm/build_module.py:
```diff
@@ -40,7 +40,7 @@ def run_process(cmd, pattern=None, env=None):
 
 @register_func
-def tvm_callback_exec_evaluate(platform, mode):
+def tvm_callback_exec_evaluate(platform, mode, empty):
```
Rename the argument. It's unclear what is "empty".

seanlatias commented on python/heterocl/tvm/build_module.py:
```diff
@@ -107,6 +107,7 @@ def tvm_callback_exec_evaluate(platform, mode):
       cmd = "cd project; " + \
             "XCL_EMULATION_MODE=sw_emu ./host build_dir" + \
             ".sw_emu." + device + "/kernel.xclbin"
+      if empty: cmd = "cd __tmp__; ./host"
```
We are not using `__tmp__` anymore, right?

zhangzhiru replied:
We absolutely need to avoid hardcoding paths or directory names in our code. We should have an API to query such information instead.

Hecmay replied:
It's fixed. We are not using any hardcoding abs path right now.

seanlatias commented on python/heterocl/schedule.py:
```diff
@@ -136,9 +164,45 @@ def reuse_at(self, target, parent, axis, name=None):
-           stream_type=_expr.StreamExpr.Channel, 
-           depth=1, name=None):
+           stream_type=_expr.StreamExpr.FIFO, 
+           depth=1, at=None, name=None):
```
Please update the parameter list below. What is `at`?

Hecmay replied:
`at` is used to specify the memory / storage media and port.

Hecmay replied:
I considered using `interface` but it is kinda to long. And the short `if` is not allowed.

zhangzhiru replied:
What does "memory/storage media & port" exactly mean?

Hecmay replied:
It means which memory the data resides in. e.g., `s.to(a, p.xcel, at=p.xcel.ddr[0])`. To move data to ddr (channel 0) attached to the accelerator.

zhangzhiru replied:
I think we should just use the term "media" instead of "at".
What is the meaning of "None" in this case?

Hecmay replied:
That's used for default mode. If it's not specified by users, HeteroCL will set it as DDR, and move the data to off-chip DRAM.

Hecmay replied:
Just discussed with Sean, to simplify the interface, we can use `s.to(tensor, p.xcel.hbm[0])` to move the data to accelerator, and have it saved in the HBM attached to that accelerator.

seanlatias commented on samples/sobel/sobel_stream.py:
```diff
@@ -0,0 +1,61 @@
+from PIL import Image
```
We should somehow add a test for this sample.

seanlatias replied:
For example, we can compare the results before and after applying `.to`.

seanlatias commented on tests/test_compute_intrinsic.py:
```diff
@@ -31,8 +31,8 @@ def test_large_int():
 
-    np_a = np.random.randint(1, 10, (10,))
-    np_b = np.random.randint(1, 10, (10,))
+    np_a = np.random.randint(1, 11, (10,))
```
This is not the correct one. The `10` is the correct one.

seanlatias commented on tvm/HalideIR/src/ir/IR.h:
```diff
@@ -1057,16 +1057,15 @@ struct KernelDef : public StmtNode<KernelDef> {
   std::string name;
-  // args to stream data 
-  Array<Expr> channels;
+  Array<Array<Expr>> channels;
```
Please don't write in this style. Use `Array<Array<Expr> >` instead.

seanlatias commented on python/heterocl/tvm/schedule.py:
```diff
@@ -334,36 +334,42 @@ def reuse_at(self, target, parent, axis, name):
 
-    def to(self, tensor, dst, src, index=0,
-           types=_expr.StreamExpr.Channel, depth=1):
+    def join(self, target, src, dst=None, types=_expr.StreamExpr.FIFO, depth=1):
+        """ join multiple writes to target tensor """
+        return _api_internal._ScheduleJoin(self, target, src, dst, types, depth)
+
+    def to(self, tensor, dst, src, axis=0,
+           types=_expr.StreamExpr.FIFO, depth=1):
```
Can we use `type` instead of `types`? Also can we have the same order for `src` and `dst`?

Hecmay replied:
Sure. But `join` is from sources to destination. Having dest before src looks weird to me.

seanlatias replied:
Then `to` should also be from source to destination?

Hecmay replied:
for data movement between host-device, it's from target to dest. for data movement between stages, yes, we should change to src to dest.

seanlatias commented on python/heterocl/schedule.py:
```diff
@@ -166,25 +233,40 @@ def to(self, tensors, dst, src=None, axis=0,
-                except AttributeError:
-                    target = tensor
+                elif isinstance(tensor, tuple):
+                    src, target = tensor
+                    src = self[src]
```
These two lines are confusing.

seanlatias commented on python/heterocl/schedule.py:
```diff
@@ -158,6 +221,10 @@ def to(self, tensors, dst, src=None, axis=0,
         depth : channel depth
             The streaming channel depth
 
+        at : memory interface
+            The memory device for the tensor to be stored
+            e.g., at=p.xcel.ddr[0] to move data to dram
+
```
Remove this.
